Added milking order c++ solution (#2108)
cpinitiative/usaco-guide: solutions/usaco-838.mdx

id: usaco-838
source: USACO Gold 2018 Open
title: Milking Order
author: Qi Wang, Melody Yu, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_gold_open18.html)

## Video Solution

Note: The video solution might not be the same as other solutions. Code in C++.

<Youtube id="OaL9vEkShyk" />

## Solution

We can binary search on the last valid value of $M$ such that the ordering can
be satisfied. For each iteration of the binary search, we only check if the
interval of times $[0, x)$ is valid. Finally, to

> find the lexicographically smallest TopoSort, we use a `priority_queue` to process

Reviewer: just say priority queue since not everyone is c++

> find the lexicographically smallest TopoSort, we use a `priority_queue` to process

Reviewer: ```suggestion
find the lexicographically smallest topological sort, we use a priority queue to process
```

interval of times $[0, x)$ is valid. Finally, to
find the lexicographically smallest TopoSort, we use a `priority_queue` to process
the ids from least to greatest.

> 

Reviewer: i would move this outside of implementation as it is an explanation

**Time Complexity:** $\mathcal{O}((E + N \log N) \log M)$

<LanguageSection>

<CPPSection>

```cpp
//CodeSnip{CPP Short Template}
//BeginCodeSnip{Topological Sort}

> // /**
> //  * Description: sorts vertices such that if there exists an edge x->y, then x goes before y
> //  * Source: KACTL
> //  * Verification: https://open.kattis.com/problems/quantumsuperposition
> //  */

Reviewer: ```suggestion
/*
  * Description: sorts vertices such that if there exists an edge x->y, then x goes before y
  * Source: KACTL
  * Verification: https://open.kattis.com/problems/quantumsuperposition
  */
```

struct TopoSort {
	int N; vi in, res;
	vector<vi> adj;
	void init(int _N) { N = _N; in.resize(N); adj.resize(N); }
	void ae(int x, int y) { adj[x].pb(y), in[y]++; }
	bool sort() {
		priority_queue<int, vi, greater<int>> todo;
		for(int i = 1; i < N; i++) if (!in[i]) todo.push(i);
		while (sz(todo)) {
			int x = todo.top(); todo.pop(); res.pb(x);
			for(const auto &i : adj[x]) if (!(--in[i])) todo.push(i);
		}
		return sz(res) == N - 1;
	}
};
//EndCodeSnip{Topological Sort}

vi ret;
vector<vi> order;
int n, m;

bool check(int x) {
	TopoSort T; T.init(n + 1);
	for(int i = 0; i < x; i++) {
		for(int j = 0; j < sz(order[i]) - 1; j++) {
			T.ae(order[i][j], order[i][j + 1]);
		}
	}
	bool ans = T.sort();
	if(ans) ret = T.res;

	return ans;
}

int main() {
	setIO("milkorder");
	cin >> n >> m;
	order.resize(m);

	for (int i = 0; i < m; i++) {
		int k; cin >> k;
		vi v(k);
		for(int j = 0; j < k; j++) cin >> v[j];
		order[i] = v;
	}

	int lo = 0, hi = m;

> 	while (lo < hi) {  // find the largest value so that f returns true

Reviewer 2: ```suggestion
	while (lo < hi) {  // find the last successful check()
```

id: usaco-838
source: USACO Gold 2018 Open
title: Milking Order
author: Qi Wang, Melody Yu, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_gold_open18.html)

## Video Solution

Note: The video solution might not be the same as other solutions. Code in C++.

<Youtube id="OaL9vEkShyk" />

## Implementation

**Time Complexity:** $\mathcal{O}((E + N \log N) \log M)$

We can binary search on the last valid value of $M$ such that the ordering can
be satisfied. For each iteration of the binary search, we only check if the
interval of times $[0, x)$ is valid. Finally, to

> find the lexigraphically smallest TopoSort, we use a `priority_queue` to process

Reviewer 3: ```suggestion
find the lexicographically smallest TopoSort, we use a `priority_queue` to process
```

> find the lexigraphically smallest TopoSort, we use a `priority_queue` to process

Reviewer 4: ```suggestion
find the lexigraphically smallest topological sort, we use a `priority_queue` to process
```

the ids from least to greatest.

<LanguageSection>

<CPPSection>

```cpp
//CodeSnip{CPP Short Template}

> //BeginCodeSnip{Topographical Sort}

Reviewer 5: Do you mean topological sort?

// /**
//  * Description: sorts vertices such that if there exists an edge x->y, then x goes before y
//  * Source: KACTL
//  * Verification: https://open.kattis.com/problems/quantumsuperposition

> //  */

Reviewer 6: 1. what's this for
2. why do you have a `//` and a `/*`

Reviewer 4: 1. it's a description of the toposort whcih is warranted
2. this is fine. what's the issue?

Reviewer 2: > this is fine. what's the issue?

it's just not necessary, as `/*...*/` already comments the code out

> //  */

Reviewer 6: 1. why's this here
2. why do you have `//` and `/*`

struct TopoSort {
	int N; vi in, res;
	vector<vi> adj;
	void init(int _N) { N = _N; in.resize(N); adj.resize(N); }

> 	void ae(int x, int y) { adj[x].pb(y), in[y] ++; }

Reviewer 3: ```suggestion
	void ae(int x, int y) { adj[x].pb(y), in[y]++; }
```

	bool sort() {
		priority_queue<int, vi, greater<int>> todo;
		for(int i = 1; i < N; i++) if (!in[i]) todo.push(i);
		while (sz(todo)) {
			int x = todo.top(); todo.pop(); res.pb(x);
			for(const auto &i : adj[x]) if (!(--in[i])) todo.push(i);
		}
		return sz(res) == N - 1;
	}
};

> //EndCodeSnip{Topographical Sort}

Reviewer 5: ```suggestion
//EndCodeSnip{Topological Sort}
```

Author: good catch

> const int mx = 2e5+1;

Reviewer 3: ```suggestion
const int mx = 2e5 + 1;
```

> const int mx = 2e5+1;

Reviewer 6: did a ctrl f, this is used literally nowhere
delete it?

vi ret;
vector<vi> order;
int n, m;

bool check(int x) {
	TopoSort T; T.init(n + 1);

> 	for(int i = 0; i < x; i++) {

Reviewer 6: put a space after the keyword please
like
```suggestion
	for (int i = 0; i < x; i++) {
```
you've done it elsewhere so just be consistent

		for(int j = 0; j < sz(order[i]) - 1; j++) {
			T.ae(order[i][j], order[i][j + 1]);
		}
	}
	bool ans = T.sort();
	if(ans) ret = T.res;

	return ans;
}

int main() {
	setIO("milkorder");
	cin >> n >> m;
	order.resize(m);

> 	for(int i = 0; i < m; i++) {

Reviewer 6: please be consistent with the braces
```suggestion
	for (int i = 0; i < m; i++) {
```

		int k; cin >> k;
		vi v(k);
		for(int j = 0; j < k; j++) cin >> v[j];
		order[i] = v;
	}

	int lo = 0, hi = m;

> 

Reviewer 6: ```suggestion
```

> 	while (lo < hi) { // find last index f is true

Reviewer 6: ```suggestion
	while (lo < hi) {  // find the largest value that has f return true
```

> 
> 	while (lo < hi) { // find last index f is true

Reviewer 6: ```suggestion
	while (lo < hi) {  // find the largest value so that f returns true
```

		int mid = lo + (hi - lo + 1) / 2;
		check(mid) ? lo = mid : hi = mid - 1;
	}

> 	for(int i = 0; i < n; i++) {

Reviewer 6: ```suggestion
	for (int i = 0; i < n; i++) {
```

> 	for(int i = 0; i < n; i++) {

Reviewer 6: ```suggestion
	for (int i = 0; i < n; i++) {
```

		cout << ret[i];

> 		if(i != n - 1) cout << " ";

Reviewer 6: ```suggestion
		if (i != n - 1) cout << " ";
```

> 		cout << ret[i];
> 		if(i != n - 1) cout << " ";

Reviewer 6: ```suggestion
		cout << ret[i] << (i != n - 1 ? " " : "");
```

Author: er i think this is less readable

Reviewer 6: you literally had a ternary in the bsearch that could possibly be considered language abuse
this is probably fine

id: usaco-838
source: USACO Gold 2018 Open
title: Milking Order
author: Timothy Gao, Qi Wang, Melody Yu
author: Timothy Gao, Qi Wang, Melody Yu, Neo Wang
---

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_gold_open18.html)

## Video Solution

Note: The video solution might not be the same as other solutions. Code in C++.

<Youtube id="OaL9vEkShyk" />

## Explanation
## Solution

> Denote each cow as a node in a graph. Let's draw a directed edge from each node $u$ to another node $v$ if $u$ must come before $v$. Note that this can be done when we process the $M$ observations. Now, notice that an ordering is valid only if the resulting graph is a DAG. Additionally, we can use a priority queue (instead of a queue) with topological sort to generate the lexicographically minimum topological ordering. It remains to find how many observations we can follow starting from the first observation, which we can binary search for.

Reviewer 2: actually, do we want to keep this?

Author: sure, not opposed
